src/graph/topo.cc

```diff
@@ -607,62 +607,92 @@ ncclResult_t ncclTopoGetLocalNet(struct ncclTopoSystem* system, int rank, int64_
+              NCCLCHECK(xmlGetAttrStr(threadblockNode, "type", &type));
+              if (rbid >= SCKL_MAX_NUM_THREAD_BLOCKS_PER_CHANNEL){
+                WARN("Too many thread blocks are requested. Max thread blocks: %d, requested: %d", SCKL_MAX_NUM_THREAD_BLOCKS_PER_CHANNEL, rbid+1);
+                return ncclInternalError;
+              }
+              if (rbid < 0){
+                WARN("rbid must be positive. rbid: %d", rbid);
+                return ncclInternalError;
+              }              
+              scklAlgo->nBlocks = std::max(comm->scklAlgo.nBlocks, rbid+1);
+              struct scklThreadBlock* sTB = &scklAlgo->scklTB[rbid];
+              sTB->nsteps = 0;
+              sTB->peer = peer;
+              if (strcmp(type, "send") == 0){
+                sTB->type = SCKL_SEND;
+              } else if (strcmp(type, "recv") == 0) {
+                sTB->type = SCKL_RECV;
+              } else {
+                WARN("type of transfer is not supported: %s", type);
+                return ncclInternalError;
+              }
+              // setting all transfers to -1 so that the ones not set are passed during runtime.
+              for (int st=0; st<SCKL_MAX_NUM_STEPS; st++){
+                sTB->transfers[st] = -1;
               }
-              for (int p=0; p<typeOfComm->nSubs; p++) {
-                struct ncclXmlNode* peer = typeOfComm->subs[p];
-                int peerId;
-                NCCLCHECK(xmlGetAttrInt(peer, "id", &peerId));
-                // SCKL generates the same scklGraph for all channels for now. This will change in the future
-                for (int c=0; c<comm->nChannels; c++){
-                  if (isRecv) {
-                    if (comm->channels[c].sGraph.nRecvPeers < SCKL_MAX_NUM_CONN){
-                      comm->channels[c].sGraph.recv[comm->channels[c].sGraph.nRecvPeers++] = peerId;
-                    } else {
-                      WARN("Too many recv connections for device %d channel %d -- connection to %d is ignored. This may cause deadlock in initialization.", rank, c, peerId);
-                    }                    
-                  } else if (isSend){
-                    if (comm->channels[c].sGraph.nSendPeers < SCKL_MAX_NUM_CONN){
-                      comm->channels[c].sGraph.send[comm->channels[c].sGraph.nSendPeers++] = peerId;
-                    } else {
-                      WARN("Too many recv connections for device %d channel %d -- connection to %d is ignored.  This may cause deadlock in initialization.", rank, c, peerId);
-                    }
+              int ntransfers = 0;
+              for (int st=0; st<threadblockNode->nSubs; st++) {
+                struct ncclXmlNode* stepNode = threadblockNode->subs[st];
+                if (strcmp(stepNode->name, "step") == 0){
+                  int s, chunkId;
+                  NCCLCHECK(xmlGetAttrInt(stepNode, "s", &s));
+                  NCCLCHECK(xmlGetAttrInt(stepNode, "chunkId", &chunkId));
```

Comment:
If we want to keep terminology consistent between sckl-core and this, `chunkId` should rather be called `addr`. In sckl-core chunks refer to unique pieces of input data and multiple chunks will have the same address if the collective is a combining one like Reduce.

Reply:
You're correct, I will fix this to be addr instead.

Reply:
In fact, I treat it like an addr